plugin.go

```diff
@@ -0,0 +1,188 @@
+
+const (
+	RootPluginName    string        = "http"
+	PluginName        string        = "sendremotefile"
+	ContentTypeKey    string        = "Content-Type"
+	ContentTypeVal    string        = "application/octet-stream"
+	xSendRemoteHeader string        = "X-Sendremotefile"
+	defaultBufferSize int           = 10 * 1024 * 1024 // 10MB chunks
+	timeout           time.Duration = 5 * time.Second
+)
+
+type Configurer interface {
+	UnmarshalKey(name string, out any) error
+	Has(name string) bool
+}
+
+type Logger interface {
+	NamedLogger(name string) *zap.Logger
+}
+
+type Plugin struct {
+	log         *zap.Logger
+	writersPool sync.Pool
+}
+
+func (p *Plugin) Init(cfg Configurer, log Logger) error {
+	const op = rrErrors.Op("sendremotefile_plugin_init")
+
+	if !cfg.Has(RootPluginName) {
+		return rrErrors.E(op, rrErrors.Disabled)
+	}
+
+	p.log = log.NamedLogger(PluginName)
+
+	p.writersPool = sync.Pool{
+		New: func() any {
+			wr := new(writer)
+			wr.code = http.StatusOK
+			wr.data = make([]byte, 0, 10)
+			wr.hdrToSend = make(map[string][]string, 2)
+			return wr
+		},
+	}
+
+	return nil
+}
+
+func (p *Plugin) Middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		rrWriter := p.getWriter()
+		defer func() {
+			p.putWriter(rrWriter)
+			_ = r.Body.Close()
+		}()
+
+		next.ServeHTTP(rrWriter, r)
+
+		// if there is no X-Sendremotefile header from the PHP worker, just return
+		if url := rrWriter.Header().Get(xSendRemoteHeader); url == "" {
+			// re-add original headers, status code and body
+			maps.Copy(w.Header(), rrWriter.hdrToSend)
+			w.WriteHeader(rrWriter.code)
+			if len(rrWriter.data) > 0 {
+				_, err := w.Write(rrWriter.data)
+				if err != nil {
+					p.log.Error("failed to write data to the response", zap.Error(err))
+				}
+			}
+
+			return
+		}
+
+		// we already checked that that header exists
+		url := rrWriter.Header().Get(xSendRemoteHeader)
+		// delete the original X-Sendremotefile header
+		rrWriter.Header().Del(xSendRemoteHeader)
+
+		if !strings.HasPrefix(url, "http") {
+			p.log.Error("header value must start with http")
+			http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
+			return
+		}
+
+		resp, err := NewClient(url, timeout).Request()
+		if err != nil {
+			var netErr net.Error
+			if errors.As(err, &netErr) && netErr.Timeout() {
+				http.Error(w, http.StatusText(http.StatusRequestTimeout), http.StatusRequestTimeout)
+				return
+			}
+
+			p.log.Error("failed to request from the upstream", zap.Error(err))
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+			return
+		}
+
+		defer func() {
+			_ = resp.Body.Close()
+		}()
+
+		if resp.StatusCode != http.StatusOK {
+			p.log.Error("invalid upstream response status code", zap.Int("status_code", resp.StatusCode))
+			http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
+			return
+		}
+
+		var buf []byte
```

Comment:
Not a mistake, but just thinking... Would it be good to use pool with []bytes? In case of pool we'd be able to have a some pool of the hot slices of bytes.

Reply:
It will definitely improve the performance :+1: 
Updated the implementation